Embroider compatibility

When trying to build the loket app under Embroider I encountered the following errors related to the editor addon:

<details>
<summary>1. Unsafe dynamic component</summary>

> Unsafe dynamic component: this.componentPath in node_modules/.embroider/rewritten-packages/@lblod/ember-rdfa-editor.0c1c90e7/utils/_private/ember-node.js

Which comes from this file: https://github.com/lblod/ember-rdfa-editor/blob/8261f1ef8a2b35dd8fc9650832ec4605c68497dd/addon/utils/_private/ember-node.ts#L151-L163

This should be solvable by passing the component class instead of the path and invoking it without the `component` helper:

```js
this.template = hbs`
  <this.component
    @getPos=this.getPos
    @node=this.node
    @updateAttribute=this.updateAttribute
    @controller=this.controller
    @view=this.view
    @selected=this.selected
    @contentDecorations=this.contentDecorations
  >
    {{#unless this.atom}}
      <EmberNode::Slot @contentDOM={{this.contentDOM}}/>
    {{/unless}}
  </this.component>`;
```
I'm not sure if this is something used outside of the addon itself, but if it is then I guess it's a breaking change 😄.
</details>

<details>
<summary>2. Ember Velcro component not found</summary>

The following error is thrown at build time:
> ERROR in ../rewritten-packages/@lblod/ember-rdfa-editor.0c1c90e7/components/toolbar/dropdown.hbs 1:0-58
Module not found: Error: Can't resolve '#embroider_compat/components/velcro' in '/Users/windvis/Development/Projects/digitaal-loket/frontend-digitaal-loket/node_modules/@lblod/ember-rdfa-editor'

I briefly looked at it and don't see anything wrong so this is probably an Embroider Bug, I'll try to find out more about this.
</details>

<details>
<summary>3. `crypto` fallback error</summary>

The addon includes some ember-auto-import configuration to provide fallbacks for node-only code. This config isn't used by Embroider, so the build fails.

```
ERROR in ../../@graphy/core.data.factory/main.js 1:15-32
Module not found: Error: Can't resolve 'crypto' in '/Users/windvis/Development/Projects/digitaal-loket/frontend-digitaal-loket/node_modules/@graphy/core.data.factory'

BREAKING CHANGE: webpack < 5 used to include polyfills for node.js core modules by default.
This is no longer the case. Verify if you need this module and configure a polyfill for it.

If you want to include a polyfill, you need to:
	- add a fallback 'resolve.fallback: { "crypto": require.resolve("crypto-browserify") }'
	- install 'crypto-browserify'
If you don't want to include a polyfill, you can use an empty module like this:
	resolve.fallback: { "crypto": false }
```

It's solvable by adding the same config in the app. I'm not sure if there is a way to handle this from the addon, but I think documenting that it's needed is a good short term solution. I noticed it only seems to complain about the `crypto` module at the moment, so I'm not sure if the other config is still needed?
</details>

--- a/.changeset/old-avocados-jam.md
+++ b/.changeset/old-avocados-jam.md
@@ -1,0 +1,25 @@
+---
+'@lblod/ember-rdfa-editor': major
+---
+
+Add Embroider "optimized" support
+
+To support the strict Embroider "optimized" preset we needed to make a breaking change. The `componentPath` property for the `createEmberNodeView` util has been replaced by a new `component` property. Instead of providing the path to the component, the component class should be passed instead.
+
+Before:
+```js
+createEmberNodeView({
+  // ... other options
+  componentPath: 'foo',
+});
+```
+
+After:
+```js
+import Foo from 'app-name/components/foo';
+
+createEmberNodeView({
+  // ... other options
+  component: Foo,
+});
+```

--- a/README.md
+++ b/README.md
@@ -20,7 +20,7 @@ ember install @lblod/ember-rdfa-editor
 ## Basic example
 
 This section includes a basic example on how to include an instance of the editor in you application.
-This addon provides the `Rdfa::RdfaEditor` component as a main entryway to add an instance of the editor.
+This addon provides the `Editor` component as a main entryway to add an instance of the editor.
 
 The following component is an example on how you can include the editor:
 

--- a/addon/components/ember-node/link.hbs
+++ b/addon/components/ember-node/link.hbs
@@ -1,4 +1,4 @@
-<Velcro @placement='bottom' @offsetOptions={{hash mainAxis=3}} as |velcro|>
+<this.Velcro @placement='bottom' @offsetOptions={{hash mainAxis=3}} as |velcro|>
   <Pill
     class='say-link'
     @icon='link'
@@ -49,4 +49,4 @@
       />
     </Pill>
   {{/if}}
-</Velcro>
+</this.Velcro>

--- a/addon/components/ember-node/link.ts
+++ b/addon/components/ember-node/link.ts
@@ -1,8 +1,11 @@
 import { action } from '@ember/object';
 import Component from '@glimmer/component';
 import { EmberNodeArgs } from '@lblod/ember-rdfa-editor/utils/ember-node';
+import { Velcro } from 'ember-velcro';
 
 export default class Link extends Component<EmberNodeArgs> {
+  Velcro = Velcro;
+
   get href() {
     return this.args.node.attrs.href as string;
   }

--- a/addon/components/plugins/text-style/color.hbs
+++ b/addon/components/plugins/text-style/color.hbs
@@ -1,6 +1,6 @@
 {{#if @controller}}
   <div class="say-toolbar__button say-toolbar__highlight-button">
-    <Velcro @placement="bottom" @offsetOptions={{hash mainAxis=6}} as |velcro|>
+    <this.Velcro @placement="bottom" @offsetOptions={{hash mainAxis=6}} as |velcro|>
       <button
         type='button'
         class='main {{if @active "is-active"}} {{if @disabled "is-disabled"}}'
@@ -33,6 +33,6 @@
           @presetColors={{@presetColors}}
         />
       {{/if}}
-    </Velcro>
+    </this.Velcro>
   </div>
 {{/if}}

--- a/addon/components/plugins/text-style/color.ts
+++ b/addon/components/plugins/text-style/color.ts
@@ -6,6 +6,7 @@ import { SayController } from '@lblod/ember-rdfa-editor';
 import { clearColor, setColor } from '@lblod/ember-rdfa-editor/plugins/color';
 import { paintCycleHappened } from '@lblod/ember-rdfa-editor/utils/_private/editor-utils';
 import { modifier } from 'ember-modifier';
+import { Velcro } from 'ember-velcro';
 
 type Args = {
   controller: SayController;
@@ -15,6 +16,7 @@ type Args = {
 export default class ColorMenu extends Component<Args> {
   dropdownButton?: HTMLElement;
   htmlSafe = htmlSafe;
+  Velcro = Velcro;
 
   setupDropdownButton = modifier(
     (element: HTMLElement) => {

--- a/addon/components/plugins/text-style/highlight.hbs
+++ b/addon/components/plugins/text-style/highlight.hbs
@@ -14,7 +14,7 @@
     />
   </button>
   <div class="say-dropdown options">
-    <Velcro @placement="bottom" @offsetOptions={{hash mainAxis=6}} as |velcro|>
+    <this.Velcro @placement="bottom" @offsetOptions={{hash mainAxis=6}} as |velcro|>
       <button
         type="button"
         class='say-dropdown__button {{if this.dropdownOpen "is-active" ""}}'
@@ -42,7 +42,7 @@
           @presetColors={{@presetColors}}
         />
       {{/if}}
-    </Velcro>
+    </this.Velcro>
   </div>
 
 </div>

--- a/addon/components/plugins/text-style/highlight.ts
+++ b/addon/components/plugins/text-style/highlight.ts
@@ -9,6 +9,7 @@ import {
 } from '@lblod/ember-rdfa-editor/plugins/highlight';
 import { paintCycleHappened } from '@lblod/ember-rdfa-editor/utils/_private/editor-utils';
 import { modifier } from 'ember-modifier';
+import { Velcro } from 'ember-velcro';
 
 type Args = {
   controller: SayController;
@@ -18,6 +19,7 @@ type Args = {
 export default class HighlightMenu extends Component<Args> {
   dropdownButton?: HTMLElement;
   htmlSafe = htmlSafe;
+  Velcro = Velcro;
 
   setupDropdownButton = modifier(
     (element: HTMLElement) => {

--- a/addon/components/responsive-toolbar.hbs
+++ b/addon/components/responsive-toolbar.hbs
@@ -5,7 +5,7 @@
       Group=(component "toolbar/group")
       Divider=(component "toolbar/divider")
       ) to="main"}}
-      <Velcro @placement="bottom-end" @offsetOptions={{hash mainAxis=10}} as |velcro|>
+      <this.Velcro @placement="bottom-end" @offsetOptions={{hash mainAxis=10}} as |velcro|>
         {{#if this.main.enableDropdown}}
           <Toolbar::Group data-ignore-resize {{velcro.hook}}>
             <Toolbar::Button 
@@ -27,7 +27,7 @@
           Divider=(component "toolbar/divider")
           ) to="main"}}
         </div>
-      </Velcro>
+      </this.Velcro>
     </div>
   {{/if}}
   {{#if (has-block 'side')}}
@@ -36,7 +36,7 @@
       Group=(component "toolbar/group")
       Divider=(component "toolbar/divider")
       ) to='side'}}
-      <Velcro @placement="bottom-end" @offsetOptions={{hash mainAxis=10}} as |velcro|>
+      <this.Velcro @placement="bottom-end" @offsetOptions={{hash mainAxis=10}} as |velcro|>
         {{#if this.side.enableDropdown}}
           <Toolbar::Group data-ignore-resize {{velcro.hook}}>
             <Toolbar::Button 
@@ -58,7 +58,7 @@
           Divider=(component "toolbar/divider")
           ) to="side"}}
         </div>
-      </Velcro>
+      </this.Velcro>
     </div>
   {{/if}}
 </div>

--- a/addon/components/responsive-toolbar.ts
+++ b/addon/components/responsive-toolbar.ts
@@ -1,6 +1,7 @@
 import { action } from '@ember/object';
 import Component from '@glimmer/component';
 import { modifier } from 'ember-modifier';
+import { Velcro } from 'ember-velcro';
 import { tracked } from 'tracked-built-ins';
 
 type ToolbarSection = {
@@ -12,6 +13,7 @@ type ToolbarSection = {
 
 export default class ResponsiveToolbar extends Component {
   toolbar?: HTMLElement;
+  Velcro = Velcro;
 
   main: ToolbarSection = tracked({
     enableDropdown: false,

--- a/addon/components/toolbar/dropdown.hbs
+++ b/addon/components/toolbar/dropdown.hbs
@@ -1,6 +1,6 @@
 
 <div class="say-dropdown" ...attributes>
-  <Velcro @placement="bottom" @offsetOptions={{hash mainAxis=6}} as |velcro|>
+  <this.Velcro @placement="bottom" @offsetOptions={{hash mainAxis=6}} as |velcro|>
     <button
       {{this.reference}}
       class='say-dropdown__button {{if this.dropdownOpen "is-active" ""}}'
@@ -37,5 +37,5 @@
         }}
       </div>
     {{/if}}
-  </Velcro>
+  </this.Velcro>
 </div>

--- a/addon/components/toolbar/dropdown.ts
+++ b/addon/components/toolbar/dropdown.ts
@@ -4,6 +4,7 @@ import { tracked } from '@glimmer/tracking';
 import { modifier } from 'ember-modifier';
 import SayController from '@lblod/ember-rdfa-editor/core/say-controller';
 import { paintCycleHappened } from '@lblod/ember-rdfa-editor/utils/_private/editor-utils';
+import { Velcro } from 'ember-velcro';
 
 type Args = {
   controller: SayController;
@@ -11,6 +12,7 @@ type Args = {
 export default class ToolbarDropdown extends Component<Args> {
   @tracked referenceElement?: Element = undefined;
   @tracked dropdownOpen = false;
+  Velcro = Velcro;
 
   reference = modifier(
     (element) => {

--- a/addon/plugins/image/nodes/image.ts
+++ b/addon/plugins/image/nodes/image.ts
@@ -1,13 +1,15 @@
+import type { ComponentLike } from '@glint/template';
 import {
   createEmberNodeSpec,
   createEmberNodeView,
   EmberNodeConfig,
 } from '@lblod/ember-rdfa-editor/utils/ember-node';
+import Image from '@lblod/ember-rdfa-editor/components/plugins/image/node';
 import { Node as PNode } from 'prosemirror-model';
 
 const emberNodeConfig: EmberNodeConfig = {
   name: 'image',
-  componentPath: 'plugins/image/node',
+  component: Image as unknown as ComponentLike,
   inline: true,
   group: 'inline',
   draggable: true,

--- a/addon/plugins/link/nodes/link.ts
+++ b/addon/plugins/link/nodes/link.ts
@@ -4,6 +4,8 @@ import {
   createEmberNodeView,
   EmberNodeConfig,
 } from '../../../utils/ember-node';
+import type { ComponentLike } from '@glint/template';
+import Link from '@lblod/ember-rdfa-editor/components/ember-node/link';
 
 type LinkOptions = {
   interactive: boolean;
@@ -15,7 +17,7 @@ const emberNodeConfig: (options: LinkOptions) => EmberNodeConfig = (
   const { interactive } = options;
   return {
     name: 'link',
-    componentPath: 'ember-node/link',
+    component: Link as unknown as ComponentLike,
     inline: true,
     group: 'inline',
     content: 'text*',

--- a/addon/utils/_private/ember-node.ts
+++ b/addon/utils/_private/ember-node.ts
@@ -29,6 +29,7 @@ import { v4 as uuidv4 } from 'uuid';
 // eslint-disable-next-line ember/no-classic-components
 import Component from '@ember/component';
 import Owner from '@ember/owner';
+import type { ComponentLike } from '@glint/template';
 import SayController from '@lblod/ember-rdfa-editor/core/say-controller';
 import { EditorState, SayView } from '@lblod/ember-rdfa-editor';
 import SayNodeSpec from '@lblod/ember-rdfa-editor/core/say-node-spec';
@@ -105,7 +106,7 @@ export function emberComponent(
   template: TemplateFactory,
   props: EmberNodeArgs & {
     atom: boolean;
-    componentPath: string;
+    component: ComponentLike<any>; // eslint-disable-line @typescript-eslint/no-explicit-any
     contentDOM?: HTMLElement;
   },
 ): { node: HTMLElement; component: EmberInlineComponent } {
@@ -147,20 +148,21 @@ class EmberNodeView implements NodeView {
     // when a node gets updated, `update()` is called.
     // We set the new node here and pass it to the component to render it.
     this.config = emberNodeConfig;
-    const { name, componentPath, atom, inline } = emberNodeConfig;
-    this.template = hbs`{{#component this.componentPath
-                          getPos=this.getPos
-                          node=this.node
-                          updateAttribute=this.updateAttribute
-                          controller=this.controller
-                          view=this.view
-                          selected=this.selected
-                          contentDecorations=this.contentDecorations
-                        }}
+    const { name, component: componentClass, atom, inline } = emberNodeConfig;
+
+    this.template = hbs`<this.component
+                          @getPos={{this.getPos}}
+                          @node={{this.node}}
+                          @updateAttribute={{this.updateAttribute}}
+                          @controller={{this.controller}}
+                          @view={{this.view}}
+                          @selected={{this.selected}}
+                          @contentDecorations={{this.contentDecorations}}
+                        >
                           {{#unless this.atom}}
-                          <EmberNode::Slot @contentDOM={{this.contentDOM}}/>
+                            <EmberNode::Slot @contentDOM={{this.contentDOM}}/>
                           {{/unless}}
-                        {{/component}}`;
+                        </this.component>`;
     this.node = pNode;
     this.contentDOM = !atom
       ? document.createElement(inline ? 'span' : 'div', {})
@@ -187,7 +189,7 @@ class EmberNodeView implements NodeView {
         },
         controller,
         contentDOM: this.contentDOM,
-        componentPath,
+        component: componentClass,
         atom,
         view,
         selected: false,
@@ -292,7 +294,7 @@ class EmberNodeView implements NodeView {
 
 export type EmberNodeConfig = {
   name: string;
-  componentPath: string;
+  component: ComponentLike<any>; // eslint-disable-line @typescript-eslint/no-explicit-any
   inline: boolean;
   group: string;
   content?: string;

--- a/package.json
+++ b/package.json
@@ -171,6 +171,7 @@
   },
   "peerDependencies": {
     "@appuniversum/ember-appuniversum": "^2.4.2",
+    "@glint/template": "^1.1.0",
     "ember-cli-sass": "^11.0.1",
     "ember-modifier": "^3.2.7"
   },

--- a/tests/dummy/app/components/sample-ember-nodes/card.ts
+++ b/tests/dummy/app/components/sample-ember-nodes/card.ts
@@ -1,0 +1,11 @@
+import templateOnlyComponent from '@ember/component/template-only';
+
+interface Signature {
+  Blocks: {
+    default: [];
+  };
+}
+
+const Card = templateOnlyComponent<Signature>();
+
+export default Card;

--- a/tests/dummy/app/dummy-nodes/card.ts
+++ b/tests/dummy/app/dummy-nodes/card.ts
@@ -1,12 +1,14 @@
+import type { ComponentLike } from '@glint/template';
 import {
   createEmberNodeSpec,
   createEmberNodeView,
   EmberNodeConfig,
 } from '@lblod/ember-rdfa-editor/utils/_private/ember-node';
+import Card from 'dummy/components/sample-ember-nodes/card';
 
 const emberNodeConfig: EmberNodeConfig = {
   name: 'card',
-  componentPath: 'sample-ember-nodes/card',
+  component: Card as unknown as ComponentLike,
   inline: false,
   group: 'block',
   content: 'block+',

--- a/tests/dummy/app/dummy-nodes/counter.ts
+++ b/tests/dummy/app/dummy-nodes/counter.ts
@@ -1,3 +1,4 @@
+import type { ComponentLike } from '@glint/template';
 import { emberApplicationPluginKey } from '@lblod/ember-rdfa-editor/plugins/ember-application';
 import {
   createEmberNodeSpec,
@@ -6,10 +7,11 @@ import {
 } from '@lblod/ember-rdfa-editor/utils/_private/ember-node';
 import { optionMapOr } from '@lblod/ember-rdfa-editor/utils/_private/option';
 import IntlService from 'ember-intl/services/intl';
+import Counter from 'dummy/components/sample-ember-nodes/counter';
 
 const emberNodeConfig: EmberNodeConfig = {
   name: 'counter',
-  componentPath: 'sample-ember-nodes/counter',
+  component: Counter as unknown as ComponentLike,
   inline: true,
   group: 'inline',
   atom: true,

--- a/tests/dummy/app/dummy-nodes/dropdown.ts
+++ b/tests/dummy/app/dummy-nodes/dropdown.ts
@@ -1,12 +1,14 @@
+import type { ComponentLike } from '@glint/template';
 import {
   createEmberNodeSpec,
   createEmberNodeView,
   EmberNodeConfig,
 } from '@lblod/ember-rdfa-editor/utils/_private/ember-node';
+import Dropdown from 'dummy/components/sample-ember-nodes/dropdown';
 
 const emberNodeConfig: EmberNodeConfig = {
   name: 'dropdown',
-  componentPath: 'sample-ember-nodes/dropdown',
+  component: Dropdown as unknown as ComponentLike,
   inline: true,
   group: 'inline',
   atom: true,

--- a/tests/test-utils.ts
+++ b/tests/test-utils.ts
@@ -58,7 +58,7 @@ export function delayMs(ms: number): Promise<void> {
  */
 export async function renderEditor() {
   await render(hbs`
-      <Rdfa::RdfaEditor
+      <Editor
               @rdfaEditorInit={{this.rdfaEditorInit}}
               @profile="default"
               class="rdfa-playground"

